# Objectives

#### **Data Overview**

- **movie.csv**: The file contains information related to the movies and their genres. All the genres contains binary values. 
 * **Columns**: 
   * **movie id:** id assigned to every movie
   * **movie title:** Title of the movie
   * **release date:** Date of release of the movie
   * The rest of the columns are different movie **genres** as follows - Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western <br>
 <br>
- **user.csv**: It contains information about the users who have rated the movies. 
 * **Columns:** 
   *  **user id:** id assigned to every user 
    * **age:** Age of the user
    * **gender:** Gender of the user
    * **occupation:** Occupation of the user
    * **zip code:** Zip code of the user<br><br>
- **ratings.csv**: It contains information for ratings given by the users to a particular movie. 
 * **Columns:** 
   * **user id:** id assigned to every user
   * **movie id:** id assigned to every movie
   * **rating:** the rating given by the user
   * **timestamp:** Time recorded when the user gave a rating<br><br>

# Importing libraries

In [1]:
import numpy as np
import pandas as pd
import datetime as dt
import warnings
warnings.filterwarnings('ignore')

# Reading dataset

In [2]:
movie = pd.read_csv(r"C:\Users\arvin\Downloads\Archive (4)\movie.csv")
rating = pd.read_csv(r"C:\Users\arvin\Downloads\Archive (4)\ratings.csv")
user = pd.read_csv(r"C:\Users\arvin\Downloads\Archive (4)\user.csv")

# Checking datatypes, Getting shapes, Gettting statistical summary

In [3]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680 entries, 0 to 1679
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   movie id      1680 non-null   int64 
 1   movie title   1680 non-null   object
 2   release date  1680 non-null   object
 3   Action        1680 non-null   int64 
 4   Adventure     1680 non-null   int64 
 5   Animation     1680 non-null   int64 
 6   Childrens     1680 non-null   int64 
 7   Comedy        1680 non-null   int64 
 8   Crime         1680 non-null   int64 
 9   Documentary   1680 non-null   int64 
 10  Drama         1680 non-null   int64 
 11  Fantasy       1680 non-null   int64 
 12  Film-Noir     1680 non-null   int64 
 13  Horror        1680 non-null   int64 
 14  Musical       1680 non-null   int64 
 15  Mystery       1680 non-null   int64 
 16  Romance       1680 non-null   int64 
 17  Sci-Fi        1680 non-null   int64 
 18  Thriller      1680 non-null   int64 
 19  War   

In [4]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user id    100000 non-null  int64
 1   movie id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


In [5]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


In [6]:
movie.shape

(1680, 21)

In [7]:
rating.shape

(100000, 4)

In [8]:
user.shape

(943, 5)

In [9]:
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

In [10]:
rating.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

In [11]:
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

In [12]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story    01-01-1995       0          0          1   
1         2   GoldenEye    01-01-1995       1          1          0   
2         3  Four Rooms    01-01-1995       0          0          0   
3         4  Get Shorty    01-01-1995       1          0          0   
4         5     Copycat    01-01-1995       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [13]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story    01-01-1995       0          0          1   
1         2   GoldenEye    01-01-1995       1          1          0   
2         3  Four Rooms    01-01-1995       0          0          0   
3         4  Get Shorty    01-01-1995       1          0          0   
4         5     Copycat    01-01-1995       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

# Converting datatype of release date column obejct to datetime

In [14]:
movie['release_date'] = pd.to_datetime(movie['release date'], format = "%d-%m-%Y")
movie.drop('release date', axis=1,inplace=True)

# Number of movies released after 1990

In [15]:
movie['Year'] = movie['release_date'].dt.year
movie.head()

movie id  movie title  Action  Adventure  Animation  Childrens  Comedy  \
0         1   Toy Story        0          0          1          1       1   
1         2   GoldenEye        1          1          0          0       0   
2         3  Four Rooms        0          0          0          0       0   
3         4  Get Shorty        1          0          0          0       1   
4         5     Copycat        0          0          0          0       0   

   Crime  Documentary  Drama  ...  Horror  Musical  Mystery  Romance  Sci-Fi  \
0      0            0      0  ...       0        0        0        0       0   
1      0            0      0  ...       0        0        0        0       0   
2      0            0      0  ...       0        0        0        0       0   
3      0            0      1  ...       0        0        0        0       0   
4      1            0      1  ...       0        0        0        0       0   

   Thriller  War  Western  release_date  Year  
0         0    0        0    1995-01-01  1995  
1         1    0        0    1995-01-01  1995  
2         1    0        0    1995-01-01  1995  
3         0    0        0    1995-01-01  1995  
4         1    0        0    1995-01-01  1995  

[5 rows x 22 columns]

In [16]:
years = movie['Year'].value_counts()
years

1996    355
1997    286
1995    220
1994    214
1993    126
       ... 
1966      2
1930      1
1931      1
1922      1
1932      1
Name: Year, Length: 70, dtype: int64

In [17]:
movie[movie['Year']>1990]['Year'].value_counts()

1996    355
1997    286
1995    220
1994    214
1993    126
1998     65
1992     37
1991     22
Name: Year, dtype: int64

## Insights
- We can observe that the highest number of movies with 355 in total have been released in 1996. Also, we can see that the count of the number of movies released in each year is increasing except for 1997 and 1998. So it can be said that, overall there is an upward trend in the number of movie released between the year 1991 and 1998.
- The top five 'years' with the highest number of movies released are '1996', '1997', '1995', '1994', and '1993'.

# Number of movies belongs to a particular genre

In [18]:
genres = movie.columns[2:20]
genres

Index(['Action', 'Adventure', 'Animation', 'Childrens', 'Comedy', 'Crime',
       'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical',
       'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [19]:
movie[genres].sum().sort_values(ascending=False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

## Insights
- Movies are distributed across 18 genres in total
- We can observe that 'Drama' genre has the highest number of movies with 725 in total followed by 'Comedy' at 505
- The top five genres with the highest number of movies are 'Drama', 'Comedy', 'Action', 'Thriller', and 'Romance'
- 'Fantasy' genre has the lowest number of movies (22 in count)

# Number of movies having more than one genre

In [20]:
movie["Number_of_Genres"] = movie.loc[:,genres].sum(axis=1)

In [21]:
movie[movie['Number_of_Genres']>1]['movie id'].count()

849

- There are 849 movies that are having more than one genre.
- Out of 1680 movies, 849 movies comprising around 50% have more than one genre and 831 movies have only single genre

# Most liked genres by users

In [22]:
df_merge = movie.merge(rating, on = 'movie id', how = 'inner')
df_merge.head()

movie id movie title  Action  Adventure  Animation  Childrens  Comedy  \
0         1  Toy Story        0          0          1          1       1   
1         1  Toy Story        0          0          1          1       1   
2         1  Toy Story        0          0          1          1       1   
3         1  Toy Story        0          0          1          1       1   
4         1  Toy Story        0          0          1          1       1   

   Crime  Documentary  Drama  ...  Sci-Fi  Thriller  War  Western  \
0      0            0      0  ...       0         0    0        0   
1      0            0      0  ...       0         0    0        0   
2      0            0      0  ...       0         0    0        0   
3      0            0      0  ...       0         0    0        0   
4      0            0      0  ...       0         0    0        0   

   release_date  Year  Number_of_Genres  user id  rating  timestamp  
0    1995-01-01  1995                 3      308       4  887736532  
1    1995-01-01  1995                 3      287       5  875334088  
2    1995-01-01  1995                 3      148       4  877019411  
3    1995-01-01  1995                 3      280       4  891700426  
4    1995-01-01  1995                 3       66       3  883601324  

[5 rows x 26 columns]

In [23]:
mean_rating = pd.DataFrame()

# Iterating through the columns of df_merge to extract and store the ratings for each genre
for i in genres:
  mean_rating[i] = np.where(df_merge[i]==1,df_merge['rating'],np.nan)

# printing the newly created dataframe
mean_rating.head()

Action  Adventure  Animation  Childrens  Comedy  Crime  Documentary  Drama  \
0     NaN        NaN        4.0        4.0     4.0    NaN          NaN    NaN   
1     NaN        NaN        5.0        5.0     5.0    NaN          NaN    NaN   
2     NaN        NaN        4.0        4.0     4.0    NaN          NaN    NaN   
3     NaN        NaN        4.0        4.0     4.0    NaN          NaN    NaN   
4     NaN        NaN        3.0        3.0     3.0    NaN          NaN    NaN   

   Fantasy  Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  \
0      NaN        NaN     NaN      NaN      NaN      NaN     NaN       NaN   
1      NaN        NaN     NaN      NaN      NaN      NaN     NaN       NaN   
2      NaN        NaN     NaN      NaN      NaN      NaN     NaN       NaN   
3      NaN        NaN     NaN      NaN      NaN      NaN     NaN       NaN   
4      NaN        NaN     NaN      NaN      NaN      NaN     NaN       NaN   

   War  Western  
0  NaN      NaN  
1  NaN      NaN  
2  NaN      NaN  
3  NaN      NaN  
4  NaN      NaN

In [24]:
mean_rating.mean(axis=0).sort_values(ascending = False)

Film-Noir      3.921523
War            3.815812
Drama          3.687379
Documentary    3.672823
Mystery        3.638132
Crime          3.632278
Romance        3.621705
Western        3.613269
Animation      3.576699
Sci-Fi         3.560723
Musical        3.521397
Thriller       3.509007
Adventure      3.503527
Action         3.480245
Comedy         3.394073
Childrens      3.353244
Horror         3.290389
Fantasy        3.215237
dtype: float64

## Insights
- 'Film-Noir' genre is having the highest average ratings with an average rating of 3.92 followed by 'War' genre with rating of 3.81. So, the majority of the population tends to like the 'Film-Noir' and 'War' movies compared to any other movies genres.
- Although, the total number of movies in 'Film-Noir' and 'Fantasy' are almost equal, 'Fantasy' genre is having the lowest average ratings with an average rating of 3.21.
- Among 18 genres, only 5 genres have received a rating below 3.5 which means around 72% of genres are rated above average

Till now we have done analyses on the genres and movies and generated some insights.

# Most preferred movies by users

Let's start by checking the top rated movies by the users. As we already know that `movie` and `ratings` datasets has the information related to the ratings of the movies.
Therefore, we can merge both the datasets on the basis of common variable that is `movie_id` and get the average ratings for all the movies. Since we have already merged these datasets before, we can call it again to get the values

In [25]:
avg_rating = df_merge.groupby(['movie title']).mean()[['rating']].reset_index() 

avg_rating.head()

movie title    rating
0  'Til There Was You   2.333333
1               1-900   2.600000
2      101 Dalmatians   2.908257
3        12 Angry Men   4.344000
4                  187  3.024390

In [27]:
avg_rating[avg_rating['rating']==1]

movie title  rating
8     3 Ninjas: High Noon At Mega Mountain      1.0
60        Amityville 1992: It's About Time      1.0
65            Amityville: A New Generation      1.0
66                   Amityville: Dollhouse      1.0
103                                 August      1.0
...                                     ...     ...
1576                 Very Natural Thing, A      1.0
1578     Vie est belle, La (Life is Rosey)      1.0
1604               Wend Kuuni (God's Gift)      1.0
1640                Woman in Question, The      1.0
1648                           Yankee Zulu      1.0

[68 rows x 2 columns]

In [26]:
avg_rating[avg_rating['rating']==5]

movie title  rating
30                                 Aiqing wansui      5.0
461   Entertaining Angels: The Dorothy Day Story      5.0
631                       Great Day in Harlem, A      5.0
942          Marlene Dietrich: Shadow and Light (     5.0
1170                                 Prefontaine      5.0
1270               Saint of Fort Washington, The      5.0
1274                          Santa with Muscles      5.0
1354                      Someone Else's America      5.0
1382                                    Star Kid      5.0
1466                     They Made Me a Criminal      5.0

## Insights
- We can observe movies like Great Day in Harlem, A, Prefontaine etc are top rated movies
- Movies like Shadow of Angels (Schatten der Engel), Power 98, etc are least rated
- There are 10 movies which have been rated as 5.0

# Relationship between demographic details of users and ratings of the movies

In [28]:
df_merge_all = df_merge.merge(user, on = 'user id', how = 'inner')

In [29]:
user['gender'].value_counts()

M    670
F    273
Name: gender, dtype: int64

In [30]:
df_merge_all.groupby('gender').mean()['rating']

gender
F    3.531510
M    3.529333
Name: rating, dtype: float64

In [31]:
df_merge_all.groupby('occupation').mean()['rating'].sort_values(ascending = False)

occupation
none             3.777778
lawyer           3.735316
doctor           3.688889
educator         3.670692
artist           3.653380
administrator    3.635646
scientist        3.611273
salesman         3.582944
programmer       3.568205
librarian        3.560781
other            3.552335
engineer         3.541473
technician       3.532097
student          3.515167
marketing        3.485641
retired          3.466750
entertainment    3.440783
writer           3.376152
executive        3.349794
homemaker        3.301003
healthcare       2.896220
Name: rating, dtype: float64

In [32]:
df_merge_all.groupby(['occupation','gender']).mean()['rating'].sort_values(ascending = False)

occupation     gender
none           M         3.876636
artist         M         3.875841
salesman       F         3.870206
administrator  F         3.781839
executive      F         3.773756
engineer       F         3.751724
lawyer         M         3.741379
educator       F         3.699132
doctor         M         3.688889
writer         F         3.663986
educator       M         3.660246
healthcare     M         3.639839
scientist      M         3.637311
none           F         3.632877
lawyer         F         3.623188
student        F         3.602879
librarian      F         3.580070
programmer     F         3.577566
               M         3.567674
other          M         3.563447
administrator  M         3.555233
technician     M         3.540477
librarian      M         3.537920
engineer       M         3.537676
other          F         3.531114
marketing      F         3.522624
homemaker      M         3.500000
student        M         3.484440
retired        M         3

## Insights
- Around 71% of users are Males while only 29% users are females
- The average rating given by Males and Females is almost equal with a value of around 3.53 and it can be said that the ratings given by Males and Females does not differ significantly.
- Users who are not working have given highest ratings as compared to users who are working professionals.
- Healthcare workers have given lowest ratings on an average
- The low ratings from the healthcare sector is majorly driven by female workers

***


# Executive Summary

- There has been an **upward trend** in the number of movies released till 1998.
- Movies are distributed across 18 genres in total
- The **top five** genres the highest number of movies are 'Drama', 'Comedy', 'Action', 'Thriller', and 'Romance'
- 
- The 'Film-Noir' genre has the highest average ratings with an average rating of 3.92, while the 'Fantasy' genre has the lowest average ratings with an average rating of 3.21
- Movies like Great Day in Harlem-A,Prefontaine, etc. are top-rated movies, while movies like Shadow of Angels (Schatten der Engel), Power 98, etc. are least rated.
- There are 334 movies in total which have been rated more than 100 times
- Around 71% of users are Males while only 29% users are **females**.
- The average rating given by Males and Females is almost equal with a value of around 3.53
- Users who are not working have rated movies higher compared to users who are working professionals
- Healthcare workers have given the lowest ratings on an average